Panic: send on closed channel

This happens randomly when using conc Pool (pool.New()).

panic: send on closed channel

goroutine 17805 [running]:
config-rd/vendor/github.com/sourcegraph/conc/pool.(*Pool).Go(0xc0005c2000, 0xc0a551e558)
	E:/go/src/config-rd/vendor/github.com/sourcegraph/conc/pool/pool.go:45 +0x113
main.doGGSN(0xc0e964b688?, 0x1)
	E:/go/src/config-rd/parser.go:63 +0xfd8
main.main.func3.1({{0xc031bbafc0, 0x10}, {0xc031bbafd0, 0x4}, {0xc03c3b36e0, 0x1a}, 0x1, 0x0, {0x0, 0x0}, ...})
	E:/go/src/config-rd/main.go:375 +0x1c5

--- a/pool/pool.go
+++ b/pool/pool.go
@@ -76,6 +76,10 @@ func (p *Pool) Wait() {
 
 	close(p.tasks)
 
+	// After Wait() returns, reset the struct so tasks will be reinitialized on
+	// next use. This better matches the behavior of sync.WaitGroup
+	defer func() { p.initOnce = sync.Once{} }()
+
 	p.handle.Wait()
 }
 

--- a/pool/pool_test.go
+++ b/pool/pool_test.go
@@ -122,6 +122,26 @@ func TestPool(t *testing.T) {
 		p := New().WithMaxGoroutines(42)
 		require.Equal(t, 42, p.MaxGoroutines())
 	})
+
+	t.Run("is reusable", func(t *testing.T) {
+		t.Parallel()
+		var count atomic.Int64
+		p := New()
+		for i := 0; i < 10; i++ {
+			p.Go(func() {
+				count.Add(1)
+			})
+		}
+		p.Wait()
+		require.Equal(t, int64(10), count.Load())
+		for i := 0; i < 10; i++ {
+			p.Go(func() {
+				count.Add(1)
+			})
+		}
+		p.Wait()
+		require.Equal(t, int64(20), count.Load())
+	})
 }
 
 func BenchmarkPool(b *testing.B) {

